Installation issue: boost fails with "classic" intel

### Steps to reproduce the issue

```console
$ spack spec -l boost cxxstd=17 %intel
Input spec
--------------------------------
boost%intel cxxstd=17

Concretized
--------------------------------
rwzzf7x  boost@1.78.0%intel@2021.5.0+atomic+chrono~clanglibcpp~container~context~coroutine+date_time~debug+exception~fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave cxxstd=17 visibility=hidden arch=linux-fedora35-broadwell
4meisx2      ^bzip2@1.0.8%intel@2021.5.0~debug~pic+shared arch=linux-fedora35-broadwell
4a6y247      ^zlib@1.2.11%intel@2021.5.0+optimize+pic+shared arch=linux-fedora35-broadwell
```


### Error message

<details><summary>Error message</summary><pre>
==> Installing boost-1.78.0-rwzzf7xc6pnmk6oprr4y6m7njmsrfrfe
==> No binary for boost-1.78.0-rwzzf7xc6pnmk6oprr4y6m7njmsrfrfe found: installing from source
==> Using cached archive: /home/ptb/gitrepos/spack/var/spack/cache/_source-cache/archive/86/8681f175d4bdb26c52222665793eef08490d7758529330f98d3b29dd0735bccc.tar.bz2
==> Ran patch() for boost
==> boost: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    './bootstrap.sh' '--prefix=/usr/local/spack/opt/spack/linux-fedora35-broadwell/intel-2021.5.0/boost-1.78.0-rwzzf7xc6pnmk6oprr4y6m7njmsrfrfe' '--with-toolset=intel-linux' '--with-libraries=math,filesystem,wave,date_time,thread,chrono,system,locale,random,exception,timer,atomic,program_options,serialization,test,iostreams,graph,log,regex' '--without-icu'

5 errors found in build log:
     14    InstalledDir: /opt/intel/oneapi/compiler/2022.0.2/linux/bin-llvm
     15    
     16    ###
     17    ###
     18    
     19    > icpx -x c++ -std=c++11 -O3 -s -static -DNDEBUG builtins.cpp class.cpp command.cpp compile.cpp constants.cpp cwd.cpp debug.cpp debugger.cpp
            execcmd.cpp execnt.cpp execunix.cpp filesys.cpp filent.cpp fileunix.cpp frames.cpp function.cpp glob.cpp hash.cpp hcache.cpp hdrmacro.cpp h
           eaders.cpp jam_strings.cpp jam.cpp jamgram.cpp lists.cpp make.cpp make1.cpp md5.cpp mem.cpp modules.cpp native.cpp object.cpp option.cpp out
           put.cpp parse.cpp pathnt.cpp pathsys.cpp pathunix.cpp regexp.cpp rules.cpp scan.cpp search.cpp startup.cpp subst.cpp sysinfo.cpp timestamp.c
           pp variable.cpp w32_getreg.cpp modules/order.cpp modules/path.cpp modules/property-set.cpp modules/regex.cpp modules/sequence.cpp modules/se
           t.cpp -o b2
  >> 20    ld: cannot find -lstdc++
  >> 21    ld: cannot find -lm
  >> 22    ld: cannot find -lc
  >> 23    clang++: error: linker command failed with exit code 1 (use -v to see invocation)
     24    > cp b2 bjam
  >> 25    cp: cannot stat 'b2': No such file or directory
     26    
     27    Failed to build B2 build engine
</pre></details>


### Information on your system

* **Spack:** 0.17.1-1455-80a6eab2db
* **Python:** 3.10.1
* **Platform:** linux-fedora35-broadwell
* **Concretizer:** clingo


### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/8189322/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/8189324/spack-build-out.txt)

Boost is using the intel-oneapi compiler `icpx` when it should just be using `icpc`.  My `compilers.yaml`

```yaml
- compiler:
    spec: intel@2021.5.0
    paths:
      cc: /opt/intel/oneapi/compiler/2022.0.2/linux/bin/intel64/icc
      cxx: /opt/intel/oneapi/compiler/2022.0.2/linux/bin/intel64/icpc
      f77: /opt/intel/oneapi/compiler/2022.0.2/linux/bin/intel64/ifort
      fc: /opt/intel/oneapi/compiler/2022.0.2/linux/bin/intel64/ifort
    flags: {}
    operating_system: fedora35
    target: x86_64
    modules: [oneapi]
    environment: {}
    extra_rpaths: []
- compiler:
    spec: oneapi@2022.0.0
    paths:
      cc: /opt/intel/oneapi/compiler/2022.0.2/linux/bin/icx
      cxx: /opt/intel/oneapi/compiler/2022.0.2/linux/bin/icpx
      f77: /opt/intel/oneapi/compiler/2022.0.2/linux/bin/ifx
      fc: /opt/intel/oneapi/compiler/2022.0.2/linux/bin/ifx
    flags: {}
    operating_system: fedora35
    target: x86_64
    modules: [oneapi]
    environment: {}
    extra_rpaths: []
```

maintainer: @hainest 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/boost/1.78-intel-linux-jam.patch
+++ b/var/spack/repos/builtin/packages/boost/1.78-intel-linux-jam.patch
@@ -1,0 +1,11 @@
+--- a/tools/build/src/tools/intel-linux.jam	2021-12-01 22:47:38.000000000 -0800
++++ b/tools/build/src/tools/intel-linux.jam	2022-05-03 13:40:41.569430070 -0700
+@@ -276,7 +276,7 @@
+ #
+ actions compile.c++.pch
+ {
+-    rm -f "$(<)" && LD_LIBRARY_PATH="$(RUN_PATH)" "$(CONFIG_COMMAND)" -x c++-header $(OPTIONS) $(USER_OPTIONS) -D$(DEFINES) -I"$(INCLUDES)" -c -pch-create "$(<)" "$(>)"
++    rm -f "$(<)" && LD_LIBRARY_PATH="$(RUN_PATH)" "$(CONFIG_COMMAND)" -x c++-header $(OPTIONS) $(USER_OPTIONS) -D$(DEFINES) -I"$(INCLUDES)" -c -Xclang -emit-pch -o "$(<)" "$(>)"
+ }
+ 
+ actions compile.fortran

--- a/var/spack/repos/builtin/packages/boost/package.py
+++ b/var/spack/repos/builtin/packages/boost/package.py
@@ -344,6 +344,9 @@ class Boost(Package):
     # See https://github.com/spack/spack/issues/28273
     patch("pthread-stack-min-fix.patch", when="@1.69.0:1.72.0")
 
+    # https://www.intel.com/content/www/us/en/developer/articles/technical/building-boost-with-oneapi.html
+    patch("1.78-intel-linux-jam.patch", when="@1.78 %oneapi")
+
     def patch(self):
         # Disable SSSE3 and AVX2 when using the NVIDIA compiler
         if self.spec.satisfies('%nvhpc'):
@@ -356,6 +359,10 @@ class Boost(Package):
 
             filter_file('-fast', '-O1', 'tools/build/src/tools/pgi.jam')
             filter_file('-fast', '-O1', 'tools/build/src/engine/build.sh')
+
+        # Fixes https://github.com/spack/spack/issues/29352
+        if self.spec.satisfies('@1.78 %intel') or self.spec.satisfies('@1.78 %oneapi'):
+            filter_file('-static', '', 'tools/build/src/engine/build.sh')
 
     def url_for_version(self, version):
         if version >= Version('1.63.0'):

